### ISSUE TITLE
Graceful shutdown in C#

### ISSUE THREAD

**Issue by vadimi** [grpc/grpc#14856]
### What version of gRPC and what language are you using?
gRPC 1.10.0
 
### What operating system (Linux, Windows, …) and version?
Linux (Ubuntu 17.10)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
.net core 2.0
 
### What did you do?
I'm trying to setup gRPC server graceful shutdown using the following code when I have a request in flight, I simulate it by using Task.Delay(10000) in my endpoint:

```
AssemblyLoadContext.Default.Unloading += context =>
{
    Console.WriteLine("Unload started");
    server?.ShutdownAsync().Wait();
    Console.WriteLine("Unload completed");
};
```
It looks like my server crashes as I only see the first log line. `echo $?` command returns `139` status code.

When I have 0 requests in flight everything is working as expected and I see both lines and `0` status code.
 
### What did you expect to see?
All requests in flight should successfully complete.
 
### What did you see instead?
 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Comment by tkarpinski** [comment]
I've also noticed this and tracked it down to https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/GrpcEnvironment.cs#L426

When SIGTERM is signaled, `Server::KillAsync` is called which executes `Server::ShutdownInternalAsync(true)`  

The call to `Server::ShutdownAsync` to perform a "clean" shutdown will throw an `InvalidOperationException` because shutdownRequested already equals `true`.

As a side note, this means there does not appear to be a way to allow existing connections to complete when a SIGTERM is signaled.

**Comment by vadimi** [comment]
@tkarpinski it is actually possible to do true graceful shutdown. I used `Mono.Posix.NETStandard` nuget package like that:

```
var signal = new UnixSignal(Signum.SIGTERM);
while (signal.WaitOne())
{
  Console.WriteLine("Unload started");
  server?.ShutdownAsync().Wait();
  Console.WriteLine("Unload completed");
  Environment.Exit(0);
}
```

It works fine and lets my requests in flight to successfully complete. The problem is it is a bit non standard and and the "official" way to handle `SIGTERM` in .net core is to capture `AssemblyLoadContext.Default.Unloading` event. But I have no idea why it's not working with gRPC and it's working just fine for web apps.

**Comment by jtattermusch** [comment]
What exactly is meant by "graceful shutdown" in the issue description? You have a server and you need to terminate it for some reason but you still want to finish handling all the inflight calls? (There seem to be better ways to do that than just using  AssemblyLoadContext.Default.Unloading IMHO - e.g. just start ShutdownAsync() whenever you need and have the main thread wait for the the ShutdownAsync() to finish). Nevertheless, what @tkarpinski  is describing might be correct, so I can take a look into that.

**Comment by jtattermusch** [comment]
@vadimi 
here's a more detailed analysis of the situation:

By default, gRPC already registers a DomainUnload or Domain.Unloading event, and the gRPC handle
shuts down all the channels and kills all the servers than haven't been shutdown yet (see https://github.com/grpc/grpc/blob/1c73e2c53ddecf3bd45f1c28d23a895fc7cd6e12/src/csharp/Grpc.Core/GrpcEnvironment.cs#L441).
https://github.com/grpc/grpc/blob/1c73e2c53ddecf3bd45f1c28d23a895fc7cd6e12/src/csharp/Grpc.Core/GrpcEnvironment.cs#L419

What happens when you shutdown with your custom AssemblyLoadContext.Default.Unloading event registers has been correctly analyzed by @tkarpinski   -   your custom handler basically collides with the default grpc handler (which does KillAsync() instead of ShutdownAsync() and thus kills all the inflight request). So a correct solution for you is to keep at least one foreground thread (presumably your main thread) running until you actually want to shutdown your application, and call `server.ShutdownAsync().Wait();` just before your last foreground thread exits.  The default grpc unload handler will then ignore the server because it's already been shutdown and everything will work as expected.

**Comment by vadimi** [comment]
thanks @jtattermusch for the explanation. The weird thing though is when I register my `AssemblyLoadContext.Default.Unloading` before I do `server.Start()` my handler runs before the one in `GrpcEnvironment` and I don't see the error @tkarpinski described at all, but `server.ShutdownAsync().Wait()` still exits my app with `139` status code.

Anyways I ended up handling `SIGTERM` using `Mono.Posix.NETStandard` in the main thread like you suggested and it works fine, all inflight requests successfully complete.

**Comment by tkarpinski** [comment]
@jtattermusch @vadimi  of course! registering the Unloading event handler before server.Start solves the clean shutdown issue and the KillAsync doesn't get called.

**Comment by vadimi** [comment]
@tkarpinski nope, unfortunately if you have requests inflight it still doesn't work (at least with grpc 1.10.0) and my requests still get cancelled, no exception is propagated to c# though. The only clean shutdown I was able to implement is by using `Mono.Posix.NETStandard` and handling `SIGTERM` this way.

**Comment by vadimi** [comment]
@jtattermusch yes, I have a simple repo with instructions in readme.md that reproduces the problem: https://github.com/vadimi/test-grpc

Basically the snippet is similar to this:
```cs
public override async Task<TestCallResponse> TestCall(TestCallRequest request, ServerCallContext context)
{
    // run kill pid_of_this_server
    await Task.Delay(10000);
    
    // no need to have listeners on these ports to reproduce the issue
    // no error here
    var channel1 = new Channel("0.0.0.0", 53000, ChannelCredentials.Insecure);
    // src/core/lib/gpr/sync_posix.cc:47: assertion failed: pthread_mutex_lock(mu) == 0 error after this line
    var channel2 = new Channel("0.0.0.0", 54000, ChannelCredentials.Insecure);

    return new TestCallResponse();
}
```
The repo also has some code to setup `SIGTERM` handling

**Comment by vadimi** [comment]
@jtattermusch I tried the same code with `1.13.0-pre1` and it looks like this issue is fixed, I don't see `pthread_mutex_lock` error anymore.

By the way, is there a chance you can make `System.Runtime.Loader.AssemblyLoadContext.Default.Unloading` ShutdownHook optional, for example by exposing `hooksRegistered` in `GrpcEnvironment` or something similar? It really complicates shutdown code quite a bit due to signal handling which .net core already has internally. Also .net core 2.1 has very nice [Generic Host Api](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-2.1) that easily allows to handle app lifecycle events including signal handling. Unfortunately it doesn't play nicely with gRPC because of that `System.Runtime.Loader.AssemblyLoadContext.Default.Unloading` event in Grpc.Core.

**Comment by chrisdunelm** [comment]
@jtattermusch I think we're seeing this cause a problem in the C# client libraries too, especially when running on Kubernetes.

The problem is that there are various times when Kubernetes shuts-down a pod, which causes gRPC to shut all channels.
However, the user program executing on the pod wishes to cleanly tidy up, which can involve making gRPC calls to services; but because gRPC has already shutdown this fails.

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/issues/2243 (slightly badly named!) contains an example of this, but I would expect this to be a moderately common pattern, especially in ASP.NET Core.
In this example, both gRPC and user-code shutdown routines (via ASP.NET Core graceful shutdown handling) both register on `System.Runtime.Loader.AssemblyLoadContext.Default.Unloading`.

As I understand it, the (graceful) pod shutdown process is: process(es) in the pod receive a SIGTERM; the pod continues running for the graceful shutdown period (defaults to 30 seconds I believe); then the process(es) are SIGKILLed if they are still running; then the pod is stopped.
Because both ASP.NET Core and gRPC both have graceful shutdown registered on `System.Runtime.Loader.AssemblyLoadContext.Default.Unloading`, it is currently impossible for user shutdown code to call any gRPC methods.

If I've got this correct, then I think this needs fixing fairly urgently.

**Comment by jtattermusch** [comment]
The shutdown hook was removed for .NET core in #14856. This should enable graceful shutdown handling when running in a Kubernetes container (and using .NET core runtime).  For .NET desktop framework and Mono, the shutdown hook currently cannot be removed (it could lead to a crash on shutdown).

The patched version was released as v1.13.1 (nugets have been uploaded).

**Comment by vadimi** [comment]
Just tried `1.13.1` - works great for my use case. thanks a lot!
